texk/ps2pk/pfb2pfa.1
.TH PFB2PFA 1 "1 Feb 2016"
.SH NAME
pfb2pfa \- convert a type1 pfb file (binary MSDOS) into a pfa (ASCII) 
.SH SYNOPSIS
pfb2pfa [-v] pfbfile [pfafile]
.SH DESCRIPTION
Program converts a binary MSDOS representation for a type1 PostScript
font into a readable ASCII version.  The MSDOS newline (\\r) is converted
into the UNIX newline (\\n).  The output is written in a file whose name
is the name that is provided on the command line or the basename of the
input file plus extension ".pfa". 
.PP
With the -v option you get some information about what the 
program is doing.
.SH AUTHOR
.nf
Piet Tutelaers
Maintained in TeX Live; please send bug reports or other correspondence
to tex-k@tug.org (http://lists.tug.org/tex-k).




ChangeLog
2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2015 => 2016/dev.
	* tardate.ac: Switch to 2015-06-12.

2015-05-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris (removed): Moved from here to doc/.
	* Makefile.am: Adapted.

2015-04-14  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README.solaris: New file from Mojca (without very long lines).
	* Makefile.am: Add the new file.

2015-04-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2015-03-07  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2015-03-07.

2015-02-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/dist_hook.am (new): Makefile fragment for dist-hook target.
	* Makefile.am: Use the fragment am/dist_hook.am.

2014-11-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-11-24.

2014-11-17  Karl Berry  <karl@tug.org>

	* am/script_links.am: Use w32_wrapper or w64_wrapper.

2014-10-21  Karl Berry  <karl@tug.org>

	* configure.ac: report CC/CXX values, with --version output
	if available.
	* Build: no need for objc any more.

2014-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-08-22.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2014 => 2015/dev.
	* tardate.ac: Switch to 2014-06-16.

2014-04-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Makefile.am: New names for README.* files.

2014-04-17  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): starting pretest, no more /dev.

2014-04-04  Karl Berry  <karl@tug.org>

	* Build (TL_MAKE,MAKE): set both in all cases.
	(buildenv.log): include pwd, and use .log instead of .txt.

2014-04-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-04-01.

2014-01-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2014-01-01.

2013-12-26  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (ACLOCAL): Ensure aclocal.m4 is up to date.

2013-10-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* doc: New directory for texlive.info & Co.
	* Makefile.am, configure.ac: Adapted.

2013-10-05  Karl Berry  <karl@tug.org>

	* Build: reorder variables and settings for clarity.
	(TL_CONF_BANNER): restore variable, default is empty string.
	(TL_POSTCONFIGURE): new variable, default true.

2013-09-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Drop ACLOCAL_AMFLAGS.
	* configure.ac: Use AC_CONFIG_MACRO_DIRS.
	Use KPSE_RECURSE_LIBS in m4/kpse-setup.m4 to recurse into all
	TeX specific libraries.

2013-09-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-09-01.

	* Makefile.am: Moved definition of cf_silent* from here ...
	* am/recurse.am: ... to here 

2013-08-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Add the make target 'texlinks'.

2013-07-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid still more 'parallel make' problems.

2013-07-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/rebuild.am: Avoid more 'parallel make' problems.

2013-07-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* am/*.am: Moved Makefile fragments from libs/am/ and tekk/am/.
	* am/rebuild.am: New Makefile fragment to rebuild a library.
	* am/reconfig.am: New Makefile fragment to reconfigure.
	* am/recurse.am: New Makefile fragment to recurse into subdirs.

	* Makefile.am, configure.ac: Use am/recurse.am.

2013-06-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Added Automake options 'tar-pax' for paths
	longer than 100 chars and 'no-dist-gzip'.
	* version.ac: Switch 2013 => 2014/dev'
	* tardate.ac: Switch to 2013-06-24.

2013-03-26  Karl Berry  <karl@tug.org>

	* Build: use $MAKE rather than $TL_MAKE for the make clean
	in the unlikely event of a local Makefile.

2013-03-21  Mojca Miklavec  <mojca.miklavec.lists@gmail.com>

	* Build: Remove GNUMAKE variable (after freetype2 as the only library
	which needed it has been fixed).

2013-03-01  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-03-01.

2013-02-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Building for Windows does not require 'ln -s'.
	* README.config: Document '--without-ln-s'.

2013-02-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (mandir, infodir): Change texmf => texmf-dist.

2013-01-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2013-01-24.
	* Makefile.am: Running texlinks in texk/texlive/tl_scripts.

2013-01-18  Karl Berry  <karl@tug.org>

	* Build: remove use of time.

2012-12-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Enable silent rules.

2012-11-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-11-16.

2012-10-24  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-10-24.

2012-08-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-08-19.

2012-08-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Running texlinks in texk/texlive/linked_scripts.

2012-07-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Build: Add TL_COMPILER_GFLAGS also to OBJCXXFLAGS.

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* version.ac: Switch 2012 => 2013/dev.

2012-06-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-06-15.

2012-05-21  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test for GNU make before creating Makefile,
	required for FreeBSD make and perhaps others.

2012-05-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* README: Update the list of prerequisites.
	* README.config: Mention configure options for biber.

2012-04-12  Karl Berry  <karl@tug.org>

	* Build (TL_COMPILER_GFLAGS): new envvar, used to override -g
	for all three compilers, with Build -g.

2012-04-03  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* tardate.ac: Switch to 2012-04-03.

2012-03-31  Karl Berry  <karl@tug.org>


texk/ps2pkm/pfb2pfa.c
/*
 * NAME
 *	pfb2pfa - convert a type1 pfb file (binary) into a pfa (ASCII) 
 * SYNOPSIS
 *	pfb2pfa [-v] pfbfile [pfafile]
 * DESCRIPTION
 *	Program converts a binary MSDOS representation for a type1 
 *	PostScript font into a readable ASCII version. The MSDOS 
 *	newline (\r) is converted into the UNIX newline (\n).
 *	The output is written in a file whose name is the name that
 *	is provided on the command line or the basename of the input
 *	file plus extension ".pfa".
 *
 *	With the -v option you get some information about what the 
 *	program is doing.
 * AUTHOR
 *	Piet Tutelaers (rcpt@urc.tue.nl)
 */

#include "basics.h"	/* basic definitions and fatal() */
#include <stdarg.h>
#include <stdio.h>
#include "filenames.h"  /* newname() */

#if defined(MSDOS) || defined(DOSISH)
#define NEWLINE '\012'
#else
#define NEWLINE '\n'
#endif

#define HEX_PER_LINE 30

int main (int argc, char *argv[])
{  unsigned int t, l, i;
   unsigned int l1, l2, l3, l4; 
   short c, done, verbose = 0;
   FILE *pfb, *pfa;
   char *pfbname, *pfaname = NULL;
   const char *myname = "pfb2pfa";

   while (--argc > 0 && (*++argv)[0] == '-') {
      done=0;
      while ((!done) && (c = *++argv[0]))  /* allow -bcK like options */
      	 switch (c) {
      	 case 'v':
      	    verbose = 1; break;
      	 default:
      	    fatal("%s: %c illegal option\n", myname, c);
      	 }
      }

   if (argc < 1) fatal("Usage: %s [-v] pfbfile [pfafile]\n", myname);
   
   pfbname = argv[0]; argc--; argv++;
   if (argc < 1) pfaname = newname(pfbname, ".pfa");
   else if (argc == 1) pfaname = argv[0];
   else fatal("Usage: %s [-v] pfbfile [pfafile]\n", myname);

   pfb = fopen(pfbname, RB);
   if (pfb == NULL) fatal("Can't open %s\n", pfbname);
   
   pfa = fopen(pfaname, WB);
   if (pfa == NULL) fatal("Can't open %s\n", pfaname);
   
   while(!feof(pfb)) {
      if (getc(pfb) != 128)
	 fatal("%s: not a pfb file.\n", pfbname);
      t = getc(pfb);
      if (verbose) printf("Type: %d, ", t);
      switch (t) {
      case 1:
         l1 = getc(pfb); l2 = getc(pfb); l3 = getc(pfb); l4 = getc(pfb); 
         l = l1 | l2 << 8 | l3 << 16 | l4 << 24;
         /* printf("%2x %2x %2x %2x -> %x\n", l1, l2, l3, l4, l); */
         if (verbose) printf(" plain text, length %d\n", l);
	 for (i=0; i < l ; i++) {
            c = getc(pfb);
            if (c == '\r') putc(NEWLINE, pfa);
            else putc(c, pfa);
	  }
         break;
      case 2:
         l1 = getc(pfb); l2 = getc(pfb); l3 = getc(pfb); l4 = getc(pfb); 
         l = l1 | l2 << 8 | l3 << 16 | l4 << 24;
         /* printf("%2x %2x %2x %2x -> %x\n", l1, l2, l3, l4, l); */
         if (verbose) printf(" binary data, length %d\n", l);
         for(i = 0; i < l ;i++) {
	    fprintf(pfa, "%02x", getc(pfb));
	    if ((i+1) % HEX_PER_LINE == 0) putc(NEWLINE, pfa);
	 }
	 putc(NEWLINE, pfa);
         break;
      case 3:
         if (verbose) printf("End of file\n");
         exit(0);
         break;
      default:
	 fatal("Unknown field type: %d\n", t);
      }
   }
   fclose(pfa); fclose(pfb);
   return 0;
}
